pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/ClassStats.java

```diff
@@ -159,14 +159,14 @@
      *
      * @return The result of the computation, or {@code Double.NaN} if it couldn't be performed
      */
-    /* default */ double computeWithResultOption(OperationMetricKey key, ASTAnyTypeDeclaration node,
+    /* default */ double computeWithResultOption(MetricKey<OperationMetric> key, ASTAnyTypeDeclaration node,
                                                  boolean force, MetricVersion version, ResultOption option) {
 
         List<ASTMethodOrConstructorDeclaration> ops = findOperations(node, false);
 
         List<Double> values = new ArrayList<>();
         for (ASTMethodOrConstructorDeclaration op : ops) {
-            if (key.supports(op)) {
+            if (key.getCalculator().supports(op)) {
```

Comment:
This is an interesting idea.  A design question... What is the purpose of having two subtypes, `ClassMetric` and `OperationMetric`, given the have identical method signatures?  Is the distinction between these metrics best expressed using classes/inheritence, or using other means (e.g. property of the Metric, say "type" with an enum {CLASS, OPERATION})?

Apologies for the odd question, I've not pulled the code and dug through it.

Also, I'm a bit confused by the `supports` methods.  In your example, it uses `N` variable, and not `Node` or `? extends Node`.  Is it intended to compile time constrain to the specific type, and dynamically determine if different instances of the type are supported?